compiler/rustc_hir_analysis/src/check/errs.rs

```diff
@@ -7,32 +7,101 @@ use rustc_span::Span;
 use crate::errors;
 
 /// Check for shared or mutable references of `static mut` inside expression
-pub fn maybe_expr_static_mut(tcx: TyCtxt<'_>, expr: hir::Expr<'_>) {
-    let span = expr.span;
-    let hir_id = expr.hir_id;
-    if let hir::ExprKind::AddrOf(borrow_kind, m, expr) = expr.kind
-        && matches!(borrow_kind, hir::BorrowKind::Ref)
-        && let Some(var) = path_if_static_mut(tcx, expr)
-    {
-        handle_static_mut_ref(tcx, span, var, span.edition().at_least_rust_2024(), m, hir_id);
+pub fn maybe_expr_static_mut(tcx: TyCtxt<'_>, expr: &hir::Expr<'_>) {
+    match expr.kind {
+        hir::ExprKind::AddrOf(borrow_kind, m, expr)
+            if matches!(borrow_kind, hir::BorrowKind::Ref) =>
+        {
+            if let Some(var) = path_if_static_mut(tcx, expr) {
+                handle_static_mut_ref(
+                    tcx,
+                    expr.span,
+                    var,
+                    expr.span.edition().at_least_rust_2024(),
+                    m,
+                    expr.hir_id,
+                    true,
+                );
+            } else if let hir::ExprKind::Field(expr, _) = expr.kind
```

Comment:
This doesn't look like it would handle nested field access like `&mut MY_STATIC.a.b.c.d`. I think it would be better to handle that in `path_if_static_mut` by just stripping away `ExprKind::Field` layers from `expr`.